### ISSUE TITLE
Explanation of how staticBump is calculated

### ISSUE THREAD

**Issue by cretz** [WebAssembly/binaryen#979]
I am writing an emulator for emscripten-emitted wasm, and some of the code is dependent upon staticBump from here. Say I were looking at a wast file, what would the algorithm look like to calculate staticBump or is this even possible from the wasm side? I was thinking it was just largest data offset + size combo, but that doesn't seem to always hold.

**Comment by kripken** [comment]
You can't calculate it from a wasm file, it's additional information - that's why it's provided as metadata in addition to the wasm.

Basically, the wasm file just contains the data, but there is additional space needed for stuff that is either not initialized (like the stack) or initialized to zero. The compiler knows that based on information in the source files, but that information is lost in the wasm. Instead, the metadata tells the rest of emscripten what to do in the JS code that sets things up.

In other words, it's not generally possible to emulate a wasm file by itself, it depends on the JS as well. The wasm is not standalone.

However, there is work towards standalone wasm, which does define everything needed in the wasm, see https://github.com/kripken/emscripten/wiki/WebAssembly-Standalone

**Comment by cretz** [comment]
Thanks. Closing.

I am essentially building the syscalls the emscripten-compiled wasm calls out to, e.g. [here](https://github.com/cretz/asmble/blob/5890a1cd7c01dc702c08e22dbe916197783ab5f5/src/main/kotlin/asmble/run/jvm/emscripten/Env.kt#L70) where I do `writev`. But I see now w/ the standalone is a much better target for this.
